api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java

```diff
@@ -0,0 +1,277 @@
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Stream;
+import org.assertj.core.api.SoftAssertions;
+import org.assertj.core.api.junit.jupiter.InjectSoftAssertions;
+import org.assertj.core.api.junit.jupiter.SoftAssertionsExtension;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.keycloak.admin.client.Keycloak;
+import org.keycloak.admin.client.resource.RealmResource;
+import org.keycloak.representations.idm.ClientRepresentation;
+import org.keycloak.representations.idm.RealmRepresentation;
+import org.keycloak.representations.idm.authorization.PolicyEnforcementMode;
+import org.keycloak.representations.idm.authorization.ResourceServerRepresentation;
+import org.projectnessie.client.http.HttpClient;
+import org.projectnessie.client.http.HttpRequest;
+import org.projectnessie.client.http.HttpResponse;
+import org.projectnessie.client.http.Status;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+
+@Testcontainers
+@ExtendWith(SoftAssertionsExtension.class)
+public class ITOAuth2Client {
+
+  @Container
+  private static final KeycloakContainer KEYCLOAK =
+      new KeycloakContainer().withFeaturesEnabled("preview", "token-exchange");
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ITOAuth2Client.class);
+
+  private static RealmResource master;
+  private static URI tokenEndpoint;
+
+  @InjectSoftAssertions private SoftAssertions soft;
+
+  @BeforeAll
+  static void setUpKeycloak() {
+    Keycloak keycloakAdmin = KEYCLOAK.getKeycloakAdminClient();
+    master = keycloakAdmin.realms().realm("master");
+    createClient("ResourceServer", false);
+    tokenEndpoint =
+        URI.create(KEYCLOAK.getAuthServerUrl() + "/realms/master/protocol/openid-connect/token");
+  }
+
+  /**
+   * This test exercises the OAuth2 client with different combinations of access token and refresh
+   * token lifespans and with/without sending the refresh token on the initial request.
+   *
+   * <p>For 15 seconds, the OAuth2 client will strive to keep the access token valid; in the
+   * meantime, another HTTP client, acting as a typical resource server, will attempt to use the
+   * access token to obtain a UMA (User Management Access) ticket from Keycloak, symbolizing the
+   * fact that Keycloak authorizes the client to access its resources.
+   */
+  @ParameterizedTest
+  @MethodSource
+  void testOAuth2Client(
```

Comment:
FYI I'm not entirely satisfied with this test that is admittedly very long (9*15 seconds). Today I'm going to try to create other tests focusing on details of the refresh cycle and maybe reduce this one to just one run of 15 seconds. (I think it's important to at least have one test exercising the full refresh cycle.)

Reply:
Yea - that's very long. Maybe one quick IT integration round plus a couple with a mocked HTTP backend?